everestrs/everestrs/src/lib.rs

```diff
@@ -104,73 +123,147 @@ pub trait GenericModule: Sync {
+            .as_ref()
+            .unwrap()
+            .upgrade()
+            .unwrap()
+            .on_ready();
     }
 
-    fn handle_command(&self, meta: &ffi::CommandMeta, json: ffi::JsonBlob) -> ffi::JsonBlob {
+    fn handle_command(&self, impl_id: &str, name: &str, json: ffi::JsonBlob) -> ffi::JsonBlob {
         let blob = self
-            .module_impl
-            .handle_command(&meta.implementation_id, &meta.name, json.deserialize())
+            .sub_impl
+            .read()
+            .unwrap()
+            .as_ref()
+            .unwrap()
+            .upgrade()
+            .unwrap()
+            .handle_command(impl_id, name, json.deserialize())
             .unwrap();
         ffi::JsonBlob::from_vec(serde_json::to_vec(&blob).unwrap())
     }
 
+    fn handle_variable(&self, impl_id: &str, name: &str, json: ffi::JsonBlob) {
+        self.sub_impl
+            .read()
+            .unwrap()
+            .as_ref()
+            .unwrap()
+            .upgrade()
+            .unwrap()
+            .handle_variable(impl_id, name, json.deserialize())
+            .unwrap();
+    }
+
+    pub fn publish_variable<T: serde::Serialize>(
+        &self,
+        impl_id: &str,
+        var_name: &str,
+        message: &T,
+    ) {
+        let blob = ffi::JsonBlob::from_vec(
+            serde_json::to_vec(&message).expect("Serialization of data cannot fail."),
+        );
+        (self.cpp_module)
+            .as_ref()
+            .unwrap()
+            .publish_variable(impl_id, var_name, blob);
+    }
+
+    pub fn call_command<T: serde::Serialize, R: serde::de::DeserializeOwned>(
+        &self,
+        impl_id: &str,
+        name: &str,
+        args: &T,
+    ) -> R {
+        let blob = ffi::JsonBlob::from_vec(
+            serde_json::to_vec(args).expect("Serialization of data cannot fail."),
+        );
+        let return_value = (self.cpp_module)
+            .as_ref()
+            .unwrap()
+            .call_command(impl_id, name, blob);
+        serde_json::from_slice(&return_value.data).unwrap()
+    }
+
     // TODO(hrapp): This function could use some error handling.
-    pub fn from_commandline<T: GenericModule + 'static>(module_impl: T) -> Self {
+    pub fn new() -> Self {
         let args: Args = argh::from_env();
-        let mut cpp_module = ffi::create_module(
+        let cpp_module = ffi::create_module(
             &args.module,
             &args.prefix.to_string_lossy(),
             &args.conf.to_string_lossy(),
         );
-        let manifest_json = cpp_module.as_mut().unwrap().initialize();
-        let manifest: schema::Manifest = manifest_json.deserialize();
-        let module = Self {
+
+        Self {
             cpp_module,
-            module_impl: Box::pin(module_impl),
-        };
+            sub_impl: RwLock::new(None),
+        }
+    }
+
+    pub fn set_subscriber(&self, sub_impl: Weak<dyn Subscriber>) {
+        *self.sub_impl.write().unwrap() = Some(sub_impl);
```

Comment:
from an api perspective, `set_subscriber` can only be called once, right?  If so, this should belong into the constructor.

Reply:
The better place would be the ctor but for our user facing api we have following dependencies 

- The publishers need a ref to the runtime
- the subscribers need a ref to the publishers
- The runtime needs a ref to the subscribers.

to brake this vicious cycle we pass the subscribers after the construction to the runtime

have a look at the core or for how it works together